src/nfa/nfa-intersection.cc

```diff
@@ -319,6 +320,20 @@ class Intersection
         }
     }
 
+    // Alternative for the above that uses SynchronizedUniverzalIterator
+    void compute_for_state_pair_using_sui()
```

Comment:
I don't like the name. Can be come up with something better? `compute_for_state_pair_efficiently`?

Or alternatively, I would merge this in the `compute_for_state_pair` and introduce parameter `use_synchronized_universal_iterator`.

Reply:
this is will be rewritten anyway in david's version of intersection, don't bother